storage/innobase/fil/fil0crypt.cc

```diff
@@ -1305,16 +1357,48 @@ static bool fil_crypt_space_needs_rotation(rotate_thread_t *state,
     key_state->key_version = crypt_data->encrypting_with_key_version;
   }
 
+  mutex_enter(&crypt_data->mutex);
+
   /* If used key_id is not found from encryption plugin we can't
   continue to rotate the tablespace */
+
   if (!crypt_data->is_key_found()) {
-    return false;
+    // We can end up here in case we try to encrypt tablespace but the key used
+    // by this tablespace is no longer in keyring. This can happen when keyring
+    // was changed or crypt_data is in version 1 and key's uuid is empty.
+    if (crypt_data->rotate_state.active_threads == 0 &&
+        crypt_data->encryption == FIL_ENCRYPTION_DEFAULT) {
+      ut_ad(
+          (crypt_data->private_version == 2 || strlen(crypt_data->uuid) == 0) &&
+          is_unenc_to_enc_rotation(*crypt_data));
+
+      crypt_data->key_found =
+          Encryption::tablespace_key_exists_or_create_new_one_if_does_not_exist(
+              crypt_data->key_id, server_uuid);
+
+      if (!crypt_data->key_found) {
+        mutex_exit(&crypt_data->mutex);
+        return false;  // failed to fetch or create the key - skip the
+                       // tablespace
+      }
+
+      key_state->key_version = 1;
+      // We assing here uuid to crypt_data key's uuid. If crypt_data
+      // does not make it to page0 (due to crash) we will do the same
+      // after the restart, because we will end up here again -
+      // encryption key will not be found.
+      memcpy(crypt_data->uuid, server_uuid, ENCRYPTION_SERVER_UUID_LEN);
```

Comment:
can this rotation thread become active at startup (when uuid is not ready).? Is it possible to assert that server_uuid is not empty here?
If uuid can be empty, how to do we fix it?

Reply:
They cannot be active at the startup. The encryption by encryption threads is on when srv_default_table_encryption == DEFAULT_TABLE_ENC_ONLINE_TO_KEYRING. srv_default_table_encryption is InnoDB's copy of default_table_encryption. It gets set in mysqld.cc by calling innodb_hton->fix_default_table_encryption(global_system_variables.default_table_encryption);
This is done immediately after innodb_hton->fix_tablespaces_empty_uuid.

Reply:
I have added a comment for this.